doc: update for NetBSD 9.1, add GUI Build Instructions

jarolrod opened the pull request:
The NetBSD build documentation has not seen any meaningful contribution since 2018. This PR intends
to update the documentation to be more informative and up to date. The main changes include:
  - Updated configuration instructions that now allow someone to successfully build 
  - New instructions on building the GUI
  - New instructions on how to support descriptor wallets
  - A table which shows dependency information

**Before/Master:** [render](https://github.com/bitcoin/bitcoin/blob/master/doc/build-netbsd.md)

**After/PR:** [render](https://github.com/bitcoin/bitcoin/blob/2d7a754ad3bbbfce581f4c4bdbbf1262fd920cba/doc/build-netbsd.md)

jarolrod committed:
doc: update for NetBSD 9.1, add GUI Build Instructions

NetBSD doc has not seen any meaningful contribution since 2018. This PR intends
to update the docs so that one can successfully build on the latest NetBSD release. It also adds
dependency information and instructions to build the GUI.

laanwj commented:
Concept ACK, I cannot test this but this is a great first contribution.

RandyMcMillan commented:
Currently testing...

RandyMcMillan commented on doc/build-netbsd.md:
```diff
@@ -1,81 +1,145 @@
+## Preparation
+
+### 1. Install Required Dependencies
+Install the required dependencies the usual way you [install software on NetBSD](https://www.netbsd.org/docs/guide/en/chap-boot.html#chap-boot-pkgsrc) - either with `pkg_add` or `pkgin`. The example commands below use `pkgin` which is usually run as `root` or via `sudo`.
```
pkg_add is the default package handler...
 - I believe it is best to use the default package handler.
- when enumerating the build and installation steps - we should not assume that the user has  sudo privileges (if possible).

I believe it is best to reformat this doc with that in mind.

An option is to add pkgin to the initial command.

`
netbsd# pkg_add pkgin autoconf automake libtool pkg-config git gmake boost libevent
`

This assumes the user has sudo privileges, is clearer on the procedure - and eliminates the need to change all the pkgin commands.

jarolrod replied:
Thanks for testing! NetBSD's official documentation [recommends](https://www.netbsd.org/docs/guide/en/chap-boot.html#chap-boot-pkgsrc) to use `pkgin` for most binary application tasks. I didn't include it as a dependency because it isn't, the user is free to use `pkg_add` instead of `pkgin`. I also link to documentation on how to install packages on NetBSD so the user can make a choice. Do you think I should explicitly add preparation instructions on installing `pkgin`?

jarolrod replied:
Addressed in 1e88f35883ac34914b1a1177244ebd7ff4977ef4. Make it clear that the examples use `pkgin`, link to documentation that shows the tool is the NetBSD recommended tool for installing binary (pre-built) packages

RandyMcMillan commented on doc/build-netbsd.md:
````diff
@@ -1,81 +1,145 @@
+
+### 2. Clone Bitcoin Repo
+Now that `git` and all the required dependencies are installed, let's clone the Bitcoin Core repository to a directory. All build scripts and commands will run from this directory.
+``` bash
````
I am using a completely clean os...

```
netbsd# git clone https://github.com/bitcoin/bitcoin
Cloning into 'bitcoin'...
error: failed to write new configuration file /root/bitcoin/.git/config.lock
warning: unable to unlink '/root/bitcoin/.git/config.lock': Operation not permitted
fatal: could not set 'core.repositoryformatversion' to '0'
netbsd# 
```

setting a better testing environment for NetBSD...

jarolrod replied:
That should be an error in the system side. First, make sure you have proper read/write access to that directory from your user. I could be wrong, but it looks like you are trying to write somewhere on root from a user. Some other troubleshooting tips: try allocating more space for the netBSD partition or manually deleting the config.lock which could have been put there by another program or process.

RandyMcMillan replied:
Here is a docker repo for testing NetBSD9.#

https://github.com/bitcoincore-dev/core-review/tree/docker-netbsd9

Works - figured others might find it useful...

RandyMcMillan commented on doc/build-netbsd.md:
````diff
@@ -1,81 +1,145 @@
 ```bash
 ./autogen.sh
-./configure --with-gui=no CPPFLAGS="-I/usr/pkg/include" \
+./configure --with-gui=no \
````
These code blocks don't copy and paste into the CL in NetBSD...
I think it is better to just use a single line format...
```bash

./configure --with-gui=no CPPFLAGS="-I/usr/pkg/include" LDFLAGS="-L/usr/pkg/lib" --with-boost-libdir=/usr/pkg/lib MAKE=gmake

```
It doesn't look as good - but it is much easier for the user to copy and paste...

RandyMcMillan replied:
While running ./configure ...
config.log is marked as a directory...
I am not sure if this is a result of copy and pasting the multiline command into the terminal...
Or if it is a bug... going to rebuild the image from scratch and run again...

![Screen Shot 2020-12-13 at 11 58 46 PM](https://user-images.githubusercontent.com/152159/102042224-4d3efc80-3d9f-11eb-8c6f-22e74aa9568c.png)

RandyMcMillan commented on doc/build-netbsd.md:
````diff
@@ -1,81 +1,145 @@
+###### Descriptor Wallet Support
+
+`sqlite3` is required to enable support for descriptor wallets. Skip if you don't intend to use descriptor wallets.
+``` bash
````
When adding code blocks - add a newline before the ticks - it is a markdown thing...
This ensures better cross platform rendering...

![Screen Shot 2020-12-13 at 11 38 28 PM](https://user-images.githubusercontent.com/152159/102041016-58dcf400-3d9c-11eb-8e99-3030988ed785.png)


###### Descriptor Wallet Support

`sqlite3` is required to enable support for descriptor wallets. Skip if you don't intend to use descriptor wallets.

``` bash
pkgin install sqlite3
```
---

jarolrod replied:
addressed in 1e88f35883ac34914b1a1177244ebd7ff4977ef4

RandyMcMillan commented on doc/build-netbsd.md:
````diff
@@ -1,81 +1,145 @@
+### 2. Clone Bitcoin Repo
+Now that `git` and all the required dependencies are installed, let's clone the Bitcoin Core repository to a directory. All build scripts and commands will run from this directory.
+``` bash
 git clone https://github.com/bitcoin/bitcoin.git
````
After more thinking on this - maybe there should be a note that states that the master HEAD isn't always stable.
Maybe this should be a git clone command to a release tag?

RandyMcMillan commented:
Just as above the config.log is being flagged as a directory

![Screen Shot 2020-12-14 at 5 59 06 PM](https://user-images.githubusercontent.com/152159/102145941-24b61180-3e36-11eb-91d3-1eacd56e709d.png)

RandyMcMillan commented:
### I think it is better if this doc:
- Assumes that it is a basic system - that has no prior configs or packages such as `pkgin`
- From that assumption - it addresses the git install issue related to
- `pkg_add -v mozilla-rootcerts` (ssl certs)
- `mozilla-rootcerts install`
- Not dependent on outside documentation
- Assumes only a basic level of understanding of the user

### With these assumptions - I believe Issues will be minimized - less support workload by community 

Related walk thru for git install issue:
https://troglobit.com/howto/netbsd-pkgsrc/

jarolrod commented:
updated 6ac875c -> 1e88f35

- Addressed @RandyMcMillan's suggestion of prepending code blocks with an empty line.
- Made it clear that the example commands use pkgin as it is the [recommended](https://www.netbsd.org/docs/guide/en/chap-boot.html#chap-boot-pkgsrc) tool for installing binary/pre-built packages

In regards to this [comment](https://github.com/bitcoin/bitcoin/pull/20610#issuecomment-746854104):
- NetBSD docs recommend to use pkgin, this git issue only occurs when using `pkg_add`. The [walk-through](https://troglobit.com/howto/netbsd-pkgsrc/) for the issue that is linked mentions that modern versions of NetBSD include `pkgin` and that the walkthrough is for bootstrapping `pkgin`

RandyMcMillan commented:
Looks good - thanks for the clarification and extra attention to detail. :) 
ACK https://github.com/bitcoin/bitcoin/commit/1e88f35883ac34914b1a1177244ebd7ff4977ef4

jarolrod commented:
updated from 1e88f35 -> 368c771 ([pr20610.02](https://github.com/jarolrod/bitcoin/tree/pr20610.02) -> [pr20610.03](https://github.com/jarolrod/bitcoin/tree/pr20610.03))

**Changes:**
- rebased due to conflict with #21942

jonatack commented on doc/build-netbsd.md:
```diff
@@ -1,81 +1,161 @@
+# NetBSD Build Guide
 
-This guide describes how to build bitcoind and command-line utilities on NetBSD.
+**Updated for NetBSD [9.1](https://netbsd.org/releases/formal-9/NetBSD-9.1.html)**
```
nit, I think the bold is unneeded here and for "required" and "optional" below. In general, I'd go light on italic and bold.

jonatack commented on doc/build-netbsd.md:
```diff
@@ -1,81 +1,161 @@
+ [git](https://pkgsrc.se/devel/git)               | Clone      | Version control system
+ [gmake](https://pkgsrc.se/devel/gmake)           | Compile    | Generate executables
+ [boost](https://pkgsrc.se/meta-pkgs/boost)       | Utility    | Library for threading, data structures, etc
+ [libevent](https://pkgsrc.se/devel/libevent)     | Networking | OS independent asynchronous networking
```
missing hyphen
```suggestion
 [libevent](https://pkgsrc.se/devel/libevent)     | Networking | OS-independent asynchronous networking
```

mahdiamer replied:
MehdiAmurirad

jonatack commented on doc/build-netbsd.md:
```diff
@@ -1,81 +1,161 @@
+
+### 1. Install Required Dependencies
+
+Install the required dependencies the usual way you [install software on NetBSD](https://www.netbsd.org/docs/guide/en/chap-boot.html#chap-boot-pkgsrc) - either with `pkg_add` or `pkgin`. The example commands below use `pkgin` which is the [recommended](https://www.netbsd.org/docs/guide/en/chap-boot.html#chap-boot-pkgsrc) way to install binary packages.
```
```suggestion
Install the required dependencies the usual way you [install software on NetBSD](https://www.netbsd.org/docs/guide/en/chap-boot.html#chap-boot-pkgsrc) -- either with `pkg_add` or `pkgin`. The example commands below use `pkgin` which is the [recommended](https://www.netbsd.org/docs/guide/en/chap-boot.html#chap-boot-pkgsrc) way to install binary packages.
```

jonatack commented on doc/build-netbsd.md:
```diff
@@ -1,81 +1,161 @@
+
+Install the required dependencies the usual way you [install software on NetBSD](https://www.netbsd.org/docs/guide/en/chap-boot.html#chap-boot-pkgsrc) - either with `pkg_add` or `pkgin`. The example commands below use `pkgin` which is the [recommended](https://www.netbsd.org/docs/guide/en/chap-boot.html#chap-boot-pkgsrc) way to install binary packages.
+
+Note: `pkgin` is usually run as `root` or via `sudo`.
```
```suggestion
Note: `pkgin` is usually run as `root` or with `sudo`.
```

jonatack commented on doc/build-netbsd.md:
```diff
@@ -1,81 +1,161 @@
 
+### 2. Clone Bitcoin Repo
+
+Now that `git` and all the required dependencies are installed, let's clone the Bitcoin Core repository to a directory. All build scripts and commands will run from this directory.
```
```suggestion
Now that `git` and the required dependencies are installed, let's clone the Bitcoin Core repository to a directory. All build scripts and commands will run from this directory.
```

jonatack commented on doc/build-netbsd.md:
```diff
@@ -1,81 +1,161 @@
+
+#### Wallet Dependencies
+
+It is not necessary to build wallet functionality to run bitcoind or the GUI. To enable legacy wallets, you must install `db4`. To enable [descriptor wallets](https://github.com/bitcoin/bitcoin/blob/master/doc/descriptors.md), `sqlite3` is required. Skip `db4` if you intend to *exclusively* use descriptor wallets
```
```suggestion
It is not necessary to build wallet functionality to run bitcoind or the GUI. To enable legacy wallets, you must install Berkeley DB, aka BDB or `db4`. To enable [descriptor wallets](https://github.com/bitcoin/bitcoin/blob/master/doc/descriptors.md), SQLite (`sqlite3`) is required. Skip `db4` if you intend to exclusively use descriptor wallets.
```

jonatack commented on doc/build-netbsd.md:
```diff
@@ -1,81 +1,161 @@
+
+###### Legacy Wallet Support
+
+`db4` is required to enable support for legacy wallets. Skip if you don't intend to use legacy wallets
```
```suggestion
BDB (`db4`) is required for legacy wallets. Skip if you don't intend to use legacy wallets.
```

jonatack commented on doc/build-netbsd.md:
```diff
@@ -1,81 +1,161 @@
-from ports, for the same reason as boost above (g++/libstd++ incompatibility).
-If you have to build it yourself, you can use [the installation script included
-in contrib/](/contrib/install_db4.sh) like so:
+`sqlite3` is required to enable support for descriptor wallets. Skip if you don't intend to use descriptor wallets.
```
```suggestion
SQLite (`sqlite3`) is required to enable support for descriptor wallets.
```

jonatack commented on doc/build-netbsd.md:
````diff
@@ -1,81 +1,161 @@
 ```
+###### qrencode
+
+The GUI can encode addresses in a QR Code. To build in QR support for the GUI, install `qrencode`. Skip if not using the GUI or don't want QR code functionality.
````
```suggestion
The GUI can encode addresses in a QR Code. To build in QR support for the GUI, install `qrencode`. Skip if not using the GUI or don't need QR code functionality.
```

jonatack commented on doc/build-netbsd.md:
```diff
@@ -1,81 +1,161 @@
+  [qrencode](https://pkgsrc.se/converters/qrencode)  | QR codes in GUI  | Generating QR codes (only needed when GUI enabled)
+  [zeromq](https://pkgsrc.se/net/zeromq)             | ZMQ notification | Allows generating ZMQ notifications (requires ZMQ version >= 4.0.0)
+  [sqlite3](https://pkgsrc.se/databases/sqlite3)     | SQLite DB        | Wallet storage (only needed when wallet enabled)
+  [python37](https://pkgsrc.se/lang/python37)        | Testing          | Python Interpreter (only needed when running the test suite)
```
According to doc/dependencies.md, the minimum required Python version is currently 3.6; if we can get away with specifying Python 3.6 and up, that might allow this doc to stay current longer.

mahdiamer replied:
Good

jarolrod replied:
lets leave this like this for now, when the minimum supported is 3.8, we can update it. What do you think?

Should be a while before the minimum is 3.8, right?

jonatack replied:
I don't have a strong opinion.

jonatack commented on doc/build-netbsd.md:
````diff
@@ -1,81 +1,161 @@
+To run the test suite (recommended), you will need to have Python 3 installed:
+
+```bash
+pkgin install python37
````
See comment above about the Python version.

jarolrod replied:
see: https://github.com/bitcoin/bitcoin/pull/20610#discussion_r672795788

jonatack commented on doc/build-netbsd.md:
```diff
@@ -1,81 +1,161 @@
+
+### 1. Configuration
+
+There are many ways to configure Bitcoin Core, here are a few common examples:
```
semi-colon or two sentences
```suggestion
There are many ways to configure Bitcoin Core. Here are a few common examples:
```

jonatack commented on doc/build-netbsd.md:
````diff
@@ -1,81 +1,161 @@
 ```bash
 gmake # use "-j N" here for N parallel jobs
-gmake check
+gmake check  # Run tests if Python 3 is available
````
```suggestion
gmake check # Run tests if Python 3 is available
```

jonatack reviewed (commented):
Approach ACK/almost ACK.  Suggestions follow, feel free to pick/choose/ignore.

jarolrod commented:
Updated from 368c771 to 2d7a754 ([pr20610.03](https://github.com/jarolrod/bitcoin/tree/pr20610.03 -> [pr20610.04](https://github.com/jarolrod/bitcoin/tree/pr20610.04)

Changes: Address @jonatack comments, thanks for the review!

jonatack commented on doc/build-netbsd.md:
```diff
@@ -1,81 +1,161 @@
+# NetBSD Build Guide
 
-This guide describes how to build bitcoind and command-line utilities on NetBSD.
+Updated for NetBSD [9.1](https://netbsd.org/releases/formal-9/NetBSD-9.1.html)
```
```suggestion
Updated for NetBSD [9.1](https://netbsd.org/releases/formal-9/NetBSD-9.1.html).
```

jonatack commented on doc/build-netbsd.md:
```diff
@@ -1,81 +1,161 @@
+
+### 1. Install Required Dependencies
+
+Install the required dependencies the usual way you [install software on NetBSD](https://www.netbsd.org/docs/guide/en/chap-boot.html#chap-boot-pkgsrc) -- either with `pkg_add` or `pkgin`. The example commands below use `pkgin` which is the [recommended](https://www.netbsd.org/docs/guide/en/chap-boot.html#chap-boot-pkgsrc) way to install binary packages.
```
Two identical links is a bit noisy/confusing; would drop one the first one.

jonatack commented on doc/build-netbsd.md:
````diff
@@ -1,81 +1,161 @@
 ```bash
 gmake # use "-j N" here for N parallel jobs
-gmake check
+gmake check # Run tests if Python 3 is available
````
s/Run/run/ or s/use/Use/

jonatack commented on doc/build-netbsd.md:
```diff
@@ -1,81 +1,161 @@
+
+#### Wallet Dependencies
+
+It is not necessary to build wallet functionality to run bitcoind or the GUI. To enable legacy wallets, you must install Berkeley DB, aka BDB or `db4`. To enable [descriptor wallets](https://github.com/bitcoin/bitcoin/blob/master/doc/descriptors.md), SQLite (`sqlite3`) is required. Skip `db4` if you intend to *exclusively* use descriptor wallets.
```
nit, would drop the unnecessary italics on "exclusively"

jonatack reviewed (commented):
utACK 368c7717d763ff247088b7a96a5720bee9cd38c9 with a few suggestions

fanquake commented:
Thanks, however I'm going to close this for now. I think #23565 is the direction we would like to go.

MarcoFalke commented:
I think this is orthogonal to https://github.com/bitcoin/bitcoin/pull/23565?
